File: 0011-container-with-most-water/0011-container-with-most-water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function(height) {
    // Init left and right pointer
    // Init maxArea to be 0
    let left = 0;
    let right = height.length - 1;
    let maxArea = 0;
    
    // while left < right :
    //  width is right - left
    //  height is min max height[left] and height[right]
    //  currentArea = width * height
    //  Compare current maxArea with currentArea
    //  Exp: We shift the lower height 
    //  while(height[left] < height[right]):
    //      left++;
    //  else:
    //      right--
    
    while(left < right) {
        const width = right - left;
        const currHeight = Math.min(height[left],height[right]);
        const currentArea = width * currHeight;
        
        maxArea = Math.max(currentArea,maxArea);
        
        if(height[left] < height[right]) {
            left++;
        } else {
            right--;
        }
    }
    
    //return maxArea;
    return maxArea;
    
    // Time complexity : O(n)
    // Space complexity : O(1)
};

File: 0011-container-with-most-water/0011-container-with-most-water.py
class Solution:
    # Time complexity : O(n)
    # Space complexity: O(1)
    def maxArea(self, height: List[int]) -> int:
        l, r = 0, len(height) - 1
        max_area = 0
        while l < r:
            curr_area = (r - l) * min(height[r],height[l])
            max_area = max(curr_area,max_area)
            if height[l] < height[r]:
                l += 1
            elif height[l] > height[r]:
                r -= 1
            else: 
                r -= 1
        return max_area
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        left,right = 0, len(height) - 1
        max_water = -1
        while left < right:
            curr_height = min(height[left],height[right])
            curr_width = right - left
            curr_water = curr_height * curr_width
            print(f"current water: {curr_water}")
            max_water = max(max_water,curr_water)
            if height[left] < height[right]:
                left += 1
            else: 
                right -= 1
        return max_water
        

        